File: ex_1/pyburger.py
import asyncio


async def burger():
    await asyncio.sleep(5)  # time it takes to make a burger
    print("burger made")


async def soda():
    await asyncio.sleep(1)  # time it takes to pour a soda
    print("soda pour")


async def order(order=[]):
    actions = []
    for item in order:
        match item:
            case "burger":
                actions.append(burger())
            case "soda":
                actions.append(soda())
            case _:
                print("invalid order")
    await asyncio.gather(*actions)


asyncio.run(order(["burger", "soda", "burger"]))
print("order complete")

File: ex_1/pyburger.rs
use tokio::task::JoinHandle;
use tokio::{
    spawn,
    time::{sleep, Duration},
};

async fn burger() -> () {
    sleep(Duration::from_millis(1000)).await;
    println!("burger made");
}

async fn soda() -> () {
    sleep(Duration::from_millis(100)).await;
    println!("soda pour");
}

async fn order(input_str: String) -> () {
    // we use string as input, it is easier to handle when we use this to PyO3
    let input: Vec<&str> = input_str.split(' ').collect();
    let mut actions: Vec<JoinHandle<()>> = Vec::new();
    for item in input {
        match item {
            "burger" => actions.push(spawn(burger())),
            "soda" => actions.push(spawn(soda())),
            _ => println!("invalid order"),
        }
    }
    for action in actions {
        let _ = action.await;
    }
    println!("order complete");
}

#[tokio::main]
async fn main() {
    order("burger soda burger".to_string()).await;
}
